src/librustc/infer/outlives/obligations.rs

```diff
@@ -151,13 +151,9 @@ impl<'cx, 'gcx, 'tcx> InferCtxt<'cx, 'gcx, 'tcx> {
         debug!("process_registered_region_obligations()");
 
         // pull out the region obligations with the given `body_id` (leaving the rest)
-        let mut my_region_obligations = Vec::with_capacity(self.region_obligations.borrow().len());
-        {
-            let mut r_o = self.region_obligations.borrow_mut();
-            for (_, obligation) in r_o.drain_filter(|(ro_body_id, _)| *ro_body_id == body_id) {
-                my_region_obligations.push(obligation);
-            }
-        }
+        let mut r_o = self.region_obligations.borrow_mut();
+        let my_region_obligations = r_o.drain_filter(|(ro_body_id, _)| *ro_body_id == body_id)
+                                       .map(|(_, obligation)| obligation);
```

Comment:
Pretty sure you need to collect this, *and not* hold that `borrow_mut` for that long, because the loop calls methods on `self` which could try to borrow again `self.region_obligations`. cc @nikomatsakis

Reply:
I wanted to collect this, but `the borrowed value didn't live long enough`.

Reply:
What did you write? I'm guessing it's just some scope thing.

Reply:
In general, `let v = { let mut x = y.borrow_mut(); x.foo().collect() };` doesn't work, you need another variable inside the block, e.g. `let v = { let mut x = y.borrow_mut(); let v = x.foo().collect(); v };`

Reply:
Ah yeah, that should work; I'll fix this shortly.